apps/emqx_gateway/src/emqx_gateway_cm_registry.erl

```diff
@@ -0,0 +1,140 @@
+-export([ init/1
+        , handle_call/3
+        , handle_cast/2
+        , handle_info/2
+        , terminate/2
+        , code_change/3
+        ]).
+
+-define(LOCK, {?MODULE, cleanup_down}).
+
+-record(channel, {chid, pid}).
+
+%% @doc Start the global channel registry.
+-spec(start_link(atom()) -> gen_server:startlink_ret()).
+start_link(Type) ->
+    gen_server:start_link(?MODULE, [Type], []).
+
+-spec tabname(atom()) -> atom().
+tabname(Type) ->
+    list_to_atom(lists:concat([emqx_gateway_, Type, '_channel_registry'])).
+
+%%--------------------------------------------------------------------
+%% APIs
+%%--------------------------------------------------------------------
+
+%% @doc Register a global channel.
+-spec register_channel(atom(), binary() | {binary(), pid()}) -> ok.
+register_channel(Type, ClientId) when is_binary(ClientId) ->
+    register_channel(Type, {ClientId, self()});
+
+register_channel(Type, {ClientId, ChanPid}) when is_binary(ClientId), is_pid(ChanPid) ->
+    ekka_mnesia:dirty_write(tabname(Type), record(ClientId, ChanPid)).
+
+%% @doc Unregister a global channel.
+-spec unregister_channel(atom(), binary() | {binary(), pid()}) -> ok.
+unregister_channel(Type, ClientId) when is_binary(ClientId) ->
+    unregister_channel(Type, {ClientId, self()});
+
+unregister_channel(Type, {ClientId, ChanPid}) when is_binary(ClientId), is_pid(ChanPid) ->
+    ekka_mnesia:dirty_delete_object(tabname(Type), record(ClientId, ChanPid)).
+
+%% @doc Lookup the global channels.
+-spec lookup_channels(atom(), binary()) -> list(pid()).
+lookup_channels(Type, ClientId) ->
+    [ChanPid || #channel{pid = ChanPid} <- mnesia:dirty_read(tabname(Type), ClientId)].
+
+record(ClientId, ChanPid) ->
+    #channel{chid = ClientId, pid = ChanPid}.
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+init([Type]) ->
+    Tab = tabname(Type),
+    ok = ekka_mnesia:create_table(Tab, [
+                {type, bag},
+                {ram_copies, [node()]},
+                {record_name, channel},
+                {attributes, record_info(fields, channel)},
+                {storage_properties, [{ets, [{read_concurrency, true},
+                                             {write_concurrency, true}]}]}]),
+    ok = ekka_mnesia:copy_table(Tab, ram_copies),
+    %%ok = ekka_rlog:wait_for_shards([?CM_SHARD], infinity),
+    ok = ekka:monitor(membership),
+    {ok, #{type => Type}}.
+
+handle_call(Req, _From, State) ->
+    logger:error("Unexpected call: ~p", [Req]),
+    {reply, ignored, State}.
+
+handle_cast(Msg, State) ->
+    logger:error("Unexpected cast: ~p", [Msg]),
+    {noreply, State}.
+
+handle_info({membership, {mnesia, down, Node}}, State = #{type := Type}) ->
+    Tab = tabname(Type),
+    global:trans({?LOCK, self()},
+                 fun() ->
+                     ekka_mnesia:transaction(fun cleanup_channels/2, [Node, Tab])
```

Comment:
The first argument should be a shard name.

Reply:
Yes, it is. But I haven't created a SHARD yet, I need to wait until your PR supports "add tabs to shard dynamically" 

I'm not sure if this will work without a shard?

Reply:
You can just put anything here for now. In the default `mnesia` mode this argument doesn't matter.

Reply:
ok. thanks for your help